### PR TITLE
enhancement(remap): Move VRL CLI into the Vector CLI

### LINKED ISSUE #5755: Migrate VRL CLI into Vector CLI
Stemming from the RFC in #5664, we should move the VRL CLI (including the REPL) into the Vector CLI, enabling users to e.g.:

```bash
vector vrl --input ./my-test-data.json
```

### PR DESCRIPTION
**Pull request by lucperkins**
This PR migrates the `vrl` CLI tool (previously in the `remap-cli` crate) under the Vector umbrella. Although there are some UX kinks in the command that we should work out over time, the modest aim of this PR is to move the logic over more or less verbatim.

## Binary size

The discussion in the RFC in #5664 called for a size comparison of the Vector binary with and without the VRL CLI. I ran two builds on a recent macOS machine:

Build | Command | Size (in bytes)
:-----|:-----------|:--------------
All features | `cargo build --release --no-default-features --features default` | 63,653,032
All features minus VRL CLI | `cargo build --release --no-default-features --features default-no-vrl-cli` | 63,261,288

The difference is gratifyingly small and, as far as I can tell, unlikely to be substantially different on other platforms. I suspect others would agree that subsuming `vrl` under Vector is worth the additional bytes.

Closes #5755

### CODE REVIEW

**Comment by binarylogic** [comment]
I agree that we should pull this into Vector, but wouldn't VRL as a standalone library still need this? Is there a way to leave support there and optionally enable it through our subcommand?

**Comment by lucperkins** [comment]
This only removes the `remap-cli` crate. `remap-lang` and `remap-functions` are unaffected and still independent.

**Comment by lucperkins** [comment]
@binarylogic Okay, I've switched back. `remap-cli` is there as before, just with some small modifications to make it exportable.

**Review by binarylogic — APPROVED** [review]
Docs look good! Curious if others think everything is in the correct place (separate of responsibilities).

**Review by JeanMertz — APPROVED** [review]
This looks good to me.

I agree with the proposed direction to _also_ keep the stand-alone CLI, as VRL itself isn't necessarily only useful within Vector itself.

I might have missed it, but have documented anywhere what the backward compatibility guarantees are for this subcommand? I'm asking because when I built the CLI, it was put together fairly quickly, without much regard for correct terminology/flag naming, etc.

I also have some longer-term goals of adding more capabilities to the CLI such as `vrl fmt`, `vrl check` and more, to format/validate/... VRL programs. Those _might_ be additive changes, but I can see us moving around some commands/flags when we get more subcommands.

So, to summarize: I'm happy to see this merged, but would like to keep the option open to make backward breaking changes to the CLI for now.

**Comment by lucperkins** [comment]
@JeanMertz That's a fair concern. Could we potentially put "unstable" changes in `remap-cli` behind feature flags?

**Comment by JeanMertz** [comment]
> That's a fair concern. Could we potentially put breaking changes in `remap-cli` behind feature flags?

We could, maybe? It would make things more complex to manage, and even if we did so, at some point we'd have to "promote" those backward breaking changes to the official version, so you'd just be delaying the breakage.

If we really want the CLI to be part of the first release, then I suggest we create an issue to "clean up/finish" the CLI as much as possible (without adding any new features), otherwise I'd propose we keep it out of the Vector CLI for now, until we're ready to ship it (we could still promote the stand-alone CLI as a "sneak peek" for those that really want to use the CLI at launch).

I'm happy to merge this though, and follow up with a clean-up PR before we ship VRL.

**Review by jszwedko — COMMENTED** [review]
This looks good to me! I agree with @JeanMertz that we should either clean it up to get into a state where we don't anticipate breaking changes to the CLI, or just leave it out of the initial release of remap.

**Comment by lucperkins** [comment]
@JeanMertz @jszwedko Here's what I propose:

* The purpose of this branch shifts from "let's just get the thing in there" to "let's get `vector vrl` shippable"
* Substantive changes to what's here already should be in separate PRs targeted to this branch
* If `vector vrl` feels ready to go by 0.12 we can merge, otherwise we can wait

Does that sound about right?

**Comment by jszwedko** [comment]
Sounds good to me, thanks @lucperkins 

A slight alternative to this would be to merge it, but don't include it in the default feature flag set so that it is not compiled in by default. This would let you merge to master earlier.

**Comment by JeanMertz** [comment]
I like that solution @jszwedko

**Review by jszwedko — APPROVED** [review]
Looks good to me, thanks @lucperkins !

**Comment by binarylogic** [comment]
> I might have missed it, but have documented anywhere what the backward compatibility guarantees are for this subcommand? I'm asking because when I built the CLI, it was put together fairly quickly, without much regard for correct terminology/flag naming, etc.

> A slight alternative to this would be to merge it, but don't include it in the default feature flag set so that it is not compiled in by default. This would let you merge to master earlier.

I feel like I'm misunderstanding something. Why are we so sensitive to breaking backward compatibility for a subcommand that is intended to be used manually for ad-hoc testing? This subcommand will be valuable for documentation and tutorials, and the cost of breaking backward compatibility is very low. Users will just have to use the new flags, right?

**Comment by lucperkins** [comment]
@binarylogic I tend to agree with you on this. `vector vrl` is a thing unto itself amongst the commands and it's not clear how changes in it would impact any workflows that aren't already centered around trial and error. The concern about it not quite being in ship shape is more pressing, I think, and I'll try to improve the documentation in the REPL and such in the coming days. I'm going to merge this for now, though, as the merge decision for these changes doesn't require any further discussion, while updating the feature flag set is trivial.

**Comment by JeanMertz** [comment]
@binarylogic @lucperkins, yeah, as I mentioned before, I'm fine getting this in, we just need to tidy things up a bit before we ship 0.12.

There's not a huge list of changes, but mainly we should do a pass over the flags to have them be consistently named. If we do end up making a breaking change down the line, that's unfortunate, but not breaking our promise as we haven't shipped 1.0 yet.
